### PR TITLE
[ci/release] Refactor release test e2e into package

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Replaces #22293

Adds a unit-tested and restructured ray_release package for running release tests.

Relevant changes in behavior:

Per default, Buildkite will wait for the wheels of the current commit to be available. Alternatively, users can a) specify a different commit hash, b) a wheels URL (which we will also wait for to be available) or c) specify a branch (or `user/branch` combination), in which case the latest available wheels will be used (e.g. if `master` is passed, behavior matches old default behavior).

The main subpackages are:

- Cluster manager: Creates cluster envs/computes, starts cluster, terminates cluster
- Command runner: Runs commands, e.g. as client command or sdk command
- File manager: Uploads/downloads files to/from session
- Reporter: Reports results (e.g. to database)

Much of the code base is unit tested, but there are probably some pieces missing.

Example build (waited for wheels to be built): https://buildkite.com/ray-project/kf-dev/builds/51#_
Wheel build: https://buildkite.com/ray-project/ray-builders-branch/builds/6023

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by simon-mo — APPROVED** [review]
Skimmed over the PR. the overall structure is great! Few questions:
- `release/release_tests.yaml`, should it be removed?
- should the old `e2e.py` users use `glue.py` now? How about `run_release_test.sh`?
- will `release/ray_release/command_runner/_wait_cluster.py` get called anywhere? Should we migrate the callers to use this as well?

**Comment by krfricke** [comment]
Thanks for taking a look, let me clarify a few things.

1) `release/release_tests.yaml` is the new release test configuration file. It will contain all release tests (similar to the CI pipeline yamls). This is following the propsal laid out here: https://docs.google.com/document/d/16IrwerYi2oJugnRf5hvzukgpJ6FAVEpB6stH_CiNMjY/edit#heading=h.4v5nvrp1pamm cc @rkooo567 
2) `run_release_test.sh` is what Buildkite will call to run (and repeat failing) release tests. Users of `e2e.py` should use `ray_release/scripts/run_release_test.py` instead
3) The `_wait_cluster.py` file will be copied over to the session automatically when using the SDK-based file manager. We can get rid of the old individual wait_cluster.py files as soon as we migrated all release tests to the new structure.

I failed to mention in the PR description that we use a new release test file - which is slightly different to the existing one. There is a conversion script in `scripts/convert_legacy_config.py` that can be used to make this easier (note that this script is not fully tested). Since nothing fundamentally changes about test execution the migration should be pretty swift and painless, but we may want to run the new pipeline a few days with e.g. the xgboost and tune tests first.

I'd like to add schema validation for the release test config file, I may do this in a follow up PR.

**Comment by krfricke** [comment]
By the way @rkooo567 I didn't look into the different schedules we currently have, yet - let me know which frequencies you need (currently multi(-daily), nightly, weekly).

**Comment by simon-mo** [comment]
@krfricke would you mind making the `release/release_tests.yaml` a bit more readable. You can either formatted it yourself, or a simple trick to replace regex `$-` with `\n-`.

**Comment by rkooo567** [comment]
Can we make it possible to switch the current e2e.py and the new implementation? In that way, we can quickly revert this. Also, can you run all nightly tests once with this new script?

**Comment by krfricke** [comment]
This PR does not replace any functionality, all scripts are separate to the existing infrastructure. Thus I propose the follwing migration plan:

1. We merge this PR. By itself it doesn't change anything.
2. We enable the new Buildkite pipeline and schedule.
3. For the test migration, we create separate PRs that disable the respective configurations in the old e2e config and enable them in the new config. This will switch them to use the new infrastructure, but those PRs are easy to revert.

Does this sound good?

File: release/BUILD [hunk 1]
```diff
@@ -55,3 +55,45 @@ py_test(
...
+    size = "small",
+    srcs = ["ray_release/tests/test_repeat.py"]
+)
+
+py_test(
+    name = "test_wheels",
+    tags = ["team:ml", "release_unit"],
+    size = "small",
+    srcs = ["ray_release/tests/test_wheels.py"]
+)
```
Comment (rkooo567) [inline release/BUILD:99]: Newline!

File: release/ray_release/cluster_manager/minimal.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    ClusterComputeCreateError,
+)
+from ray_release.logger import logger
+from ray_release.cluster_manager.cluster_manager import ClusterManager
+from ray_release.util import format_link, anyscale_cluster_env_build_url
+
+REPORT_S = 30.0
+
+
+class MinimalClusterManager(ClusterManager):
```
Comment (rkooo567) [inline release/ray_release/cluster_manager/minimal.py:16]: Why do we need this? Can we just combine it to a single class?

File: release/ray_release/file_manager/file_manager.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import abc
+from typing import Optional
+
+from ray_release.cluster_manager.cluster_manager import ClusterManager
+
+RAY_REMOTE_HOME = "/home/ray"
+
+
+class FileManager(abc.ABC):
```
Comment (rkooo567) [inline release/ray_release/file_manager/file_manager.py:9]: This seems to be not used? Remove it?

**Review by rkooo567 — APPROVED** [review]
LGTM. Just a few minor comments. I think some of classes are a bit redundant or not necessary?

File: release/ray_release/file_manager/file_manager.py (hunk 3 above)
Comment (krfricke) [reply release/ray_release/file_manager/file_manager.py:7]: I've removed the ray remote home variables (unneeded). The FileManager is the abstract base class of the other file managers so we should keep it as a general data transfer interface

**Comment by krfricke** [comment]
For e2e.py, fixes can be made in the master branch. For the new infra, fixes will need to be cherry picked onto the release branch.
While migrating, both will be used, after migration, we can get rid of e2e.py

**Comment by mwtian** [comment]
Sounds good.

**Pull request merged**